lib/cn/csched_sp3_work.c

```diff
@@ -287,16 +321,52 @@ sp3_work_wtype_split(
     enum cn_rule             *rule)
 {
     struct cn_tree_node *tn = spn2tn(spn);
+    struct cn_tree *tree = tn->tn_tree;
     struct kvset_list_entry *le;
     struct list_head *head;
+    uint64_t now;
+    uint kvsets;
 
     head = &tn->tn_kvset_list;
     *mark = list_last_entry_or_null(head, typeof(*le), le_link);
+    kvsets = cn_ns_kvsets(&tn->tn_ns);
 
     *action = CN_ACTION_SPLIT;
     *rule = CN_RULE_SPLIT;
 
-    return cn_ns_kvsets(&tn->tn_ns);
+    mutex_lock(&tn->tn_ss_lock);
+    now = get_time_ns();
+
+    if (!tn->tn_ss_splitting) {
+        if (tree->ct_split_cnt >= thresh->split_cnt_max || now < tree->ct_split_dly) {
+            tn->tn_ss_visits = 0;
+            kvsets = 0;
+            ev_debug(1);
+        } else if (cn_ns_kvsets(&tree->ct_root->tn_ns) >= thresh->rspill_runlen_min * 3) {
+            tn->tn_ss_visits = 0;
+            kvsets = 0;
+            ev_debug(1);
+        } else if (tn->tn_ss_spilling && tn->tn_ss_visits < thresh->split_cnt_max) {
```

Comment:
We don't want to delay split if tn->tn_ss_visits >= thresh->split_cnt_max. However, the check at Line 365 seem to override this, not sure...